### PR TITLE
containerd resource binaries

### PR DESCRIPTION
**Pull request by addyess**
In preparation for the Charmed Kubernetes 1.26 release, the `containerd` deb from Canonical was pinned on to 1.5.9 on jammy, focal, and bionic. 

In order to have more current compatible binaries for 1.26 kubelet, [LP#1996534](https://bugs.launchpad.net/ubuntu/+source/containerd/+bug/1996534) was opened, but will not be available in time for the release of the charm.  For the initial ck-1.26 the containerd will come with a 1.6.10 release of containerd resource already attached to the charm which supports amd64 and arm64

The charm is being extended to support installing binaries from the upstream [project's releases](https://github.com/containerd/containerd/releases).  A user can download the github release (say 1.6.10) for the target architecture of their cluster and attach it to the charm with:

```bash
juju attach-resource containerd containerd=/path/to/local/containerd-1.6.10-linux-amd64.tar.gz
```

Which will unpack the tar.gz and install the binaries over the existing deb packages.  To remove the resource and restore from the deb sources, one must attach an empty resource.

```bash
touch empty.tgz
juju attach-resource containerd containerd=empty.tgz
```

It is also possible to build a multiarch resource -- see this PR's [build-resources.sh] to see the process.

### CODE REVIEW

File: tests/data/charm.yaml [hunk 1]
```diff
@@ -3,10 +3,19 @@ applications:
...
     charm: docker-registry
     channel: edge
     num_units: 1
+  ##  TEMPORARY FIX FOR PASSING INTEGRATION TESTS
+  kubernetes-control-plane:
+    options:
+      enable-metrics: false
+      controller-manager-extra-args: |-
+        requestheader-client-ca-file=/root/cdk/ca.crt requestheader-allowed-names=system:kube-controller-manager,client
+  ################################################
```
Comment (addyess) [inline tests/data/charm.yaml:18]: Addressed partially by:
https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/pull/257

File: lib/charms/layer/containerd.py [hunk 2]
```diff
@@ -1,4 +1,68 @@
...
+        return ResourceFailure("Incomplete containerd resource")
+    check_call(["tar", "xfz", archive, "-C", unpack_path])
+    return _collect_resource_bins(unpack_path)
+
+
+def _collect_resource_bins(unpack_path):
+    for arch_based in unpack_path.glob("./*.tar.gz"):
+        if arch() in arch_based.name:
+            unpack_path = unpack_path / arch()
+            return _unpack_archive(arch_based, unpack_path)
```
Comment (addyess) [inline lib/charms/layer/containerd.py:53]: check to see if the original archive contains `tar.gz` files within it -- if so it's likely they are the arch based resources

Comment (addyess) [inline lib/charms/layer/containerd.py:53]: check to see if the original archive contains `tar.gz` files within it -- if so it's likely they are the arch based resources.  With the first archive containing this machine's `arch()` -- treat it like it contains the binaries

File: lib/charms/layer/containerd.py [hunk 3]
```diff
@@ -1,4 +1,68 @@
...
+    unpack_path = Path(charm_dir, "resources", "containerd")
+    return _unpack_archive(archive, unpack_path)
+
+
+def _unpack_archive(archive, unpack_path):
+    unpack_path.mkdir(exist_ok=True, parents=True)
+    archive = Path(archive)
+    filesize = archive.stat().st_size
+    if filesize == 0:
+        return None
```
Comment (addyess) [inline lib/charms/layer/containerd.py:42]: providing an empty file for the archive signifies we should use the deb source

File: reactive/containerd.py [hunk 4]
```diff
@@ -421,6 +421,12 @@ def upgrade_charm():
     # Prevent containerd apt pkg from being implicitly updated.
     apt_hold(CONTAINERD_PACKAGE)
 
+    clear_flag("containerd.resource.installed")
+    if DB.get("containerd-resource-overridden"):
+        # if a resource is currently overriding the deb,
+        # upgrade containerd from the apt packages
+        reinstall_containerd()
+
```
Comment (addyess) [inline reactive/containerd.py:429]: charm-upgrade fires when you upgrade the charm (naturally) or attach a resource.  So if the user attaches a new containerd resource, we should clear the `containerd.resource.installed` flag.  If the resource was PREVIOUSLY overriding the deb sources, we'll need to first install the deb sources and then react to any other flag changes.

File: reactive/containerd.py [hunk 5]
```diff
@@ -465,15 +471,44 @@ def install_containerd():
...
+def reinstall_containerd():
+    """Install and hold containerd with apt."""
+    apt_update(fatal=True)
+    apt_unhold(CONTAINERD_PACKAGE)
+    apt_install([CONTAINERD_PACKAGE, "--reinstall"], fatal=True)
+    apt_hold(CONTAINERD_PACKAGE)
     set_state("containerd.installed")
-    config_changed()
+    clear_flag("containerd.resource.installed")
```
Comment (addyess) [inline reactive/containerd.py:485]: this method could be called from the action `upgrade-packages` -- in which case we may need to re-apply any resource afterwards.

File: reactive/containerd.py [hunk 6]
```diff
@@ -486,7 +521,7 @@ def publish_version_to_juju():
         return
 
     output = output.decode()
-    ver_re = re.compile(r"\s*Version:\s+([\d\.]+)")
+    ver_re = re.compile(r"\s*Version:\s+v{0,1}([\d\.]+)")
```
Comment (addyess) [inline reactive/containerd.py:524]: well, so the versions built from upstream contain a `v`.  And the deb sources don't?  whatever -- handle that

**Review by stonepreston — APPROVED** [review]
LGTM 🚀

**Pull request merged**